records/2023-01-26.bounce-rate-automated-emergency-brake.md

```diff
@@ -31,47 +33,99 @@ AWS calculates the bounce rate as the number of hard bounces per the total numbe
+group by 
+    service_id
+having
+    count(id) >= 1000 -- this value is the `[bounce-rate-warning-notification-volume-minimum]`
+)
+
+select *, 100*(hard_bounces::decimal/total_emails::decimal) as bounce_rate from data
+```
 
 ## Additional considerations
 
-### How GC Notify handles bounce notifications from SES
+### Staggered roll-out
+To prevent suspending services that may not actually be causing issues with the overall bounce rate, it is suggested to roll out this feature in stages.  The first stage would include tracking bounces, calculating bounce rates, and displaying them on service dashboards. It would also include providing bounce rate guidance for services. Once the calculations have been confirmed to work satisfactorily with live data, the second stage can be rolled out which will include warning and suspending services.
+
+### How GC Notify handles asynchronous bounce notifications from SES
 GC Notify uses a lambda function named `ses-to-sqs-email-callbacks` to handle bounce notifications sent back from SES.  When a callback comes in, the lambda function parses the bounce notification and sends it a queue to be picked up and run by the celery task `process-ses-result`.  The `process-ses-result` task updates the status of the notification in the database including whether it was a hard or a soft bounce.
 
+### How GC Notify handles synchronous bounce notifications from SES
+GC Notify sends emails via the `send_email()` method in `clients/email/aws_ses.py`.  
+- :warning: There does not appear to be code to record synchronous bounces in this method.
+
+## Proposed values
+
+The values below can be used as a starting point for this feature.  They should be reviewed and adjusted as needed as the feature is tested and rolled out.
+
+| Parameter                                       | Value    | Notes                                                                                                                        |
+|-------------------------------------------------|----------|------------------------------------------------------------------------------------------------------------------------------|
+| bounce-rate-warning-percentage                  |       5% | AWS indicates a 5% bounce rate as under review, so we should warn no higher than this                                        |
+| bounce-rate-critical-percentage                 |       9% | AWS indicates a 10% or higher bounce rate as at risk, so we should suspend before this                                       |
+| bounce-rate-window                              | 24 hours | This should be at a minimum 24 hours, but could be a longer period                                                           |
+| bounce-rate-warning-notification-volume-minimum |     2500 | This should be a significantly high number to ensure we only notify users with enough volume to sway the overall bounce rate |
```

Comment:
I'd shoot with 1,000 but 2,500 could be fine too.

Reply:
Either value works! As long as `bounce-rate-warning-notification-volume-minimum` isn't less than `bounce-rate-display-volume-minimum` because in that case we would be warning people via but not showing it to them on their dashboard!

I thought I had understood one of the concerns here was notifying many low-volume services and having increased support calls because of it.  The idea was that this value is a lever we can use to reduce those support calls.  

That said, looking at the data over the past few days, a value of 1000 would mean watching 11 services vs. a value of 2500 means watching 6; not a huge difference.  I'll update this as per your recommendation, and we can always tweak this as more services onboard.

Reply:
Yes you are right, we would target more services and these would get into our support/education scope, hence more work. If we select a value that is too high though, the risk is that a high bounce rate service excluded from a high individual service limit could influence the global bounce rate. The 2,500 value seems low enough to avoid that, but I like to be extra safe with `1,000` on that front.. but I'd be fine with 2,500 too! It's a type 2 decisions anyway so we'd be fine.